p4-16/psa/PSA.mdk

```diff
@@ -2572,3 +2543,139 @@ Certainly if there are long durations of time when egress decides to
+use separate queues for multicast traffic versus unicast.
+
+
+We give some motivations for these recommendations below.
+
+1. Expectations of hosts
+
+While the Internet Protocol does not have strong ordering requirements
+for sequences of packets sent by one host to another, there are still
+widely deployed implementations of TCP that lead to significantly
+degraded throughput when the network frequently delivers packets to
+the receiver in a different order than they were transmitted.  While
+significant research and development has been done to mitigate this
+issue, e.g. later Linux TCP implementations (since 2011 when version
+2.6.35 was released) are much more resistant to this problem than
+earlier versions, there are still many commonly deployed TCP
+implementations that suffer from this issue.  See references within
+Kandula et al's work[^Kandula2007] for some of the research done
+towards making TCP more robust in the face of network packet
+reordering.
+
+Such TCP implementations interpret acknowledgements with repeated
+cumulative acknowledgement sequence numbers as a likely indication of
+packet loss in the network, and reduce their sending window in an
+effort to avoid network congestion.
+
+While applications using UDP should also be aware of possible packet
+reordering in a network, some of them behave poorly if this reordering
+becomes common[^LG2002].
+
+These expectations of hosts are a significant reason why ECMP (Equal
+Cost Multi Path) path selection and LAG (Link Aggregation Group) link
+selection are so often done by using a hash of packet header fields
+such as IP source and destination address, and TCP or UDP source and
+destination port.  Choosing among parallel paths in this way helps to
+preserve the order of packets in the same application flow, at the
+cost of not balancing the load as evenly as possible.  If a network
+device's internal implementation reordered packets, it would be an
+independent source of network reordering.
+
+[^Kandula2007]: S. Kandula, D. Katabi, S. Sinha, and A. Berger, "Dynamic load balancing without packet reordering", ACM SIGCOMM Computer Communication Review, Vol. 37, No. 2, April 2007
+[^LG2002]: M. Laor and L. Gendel, "The effect of packet reordering in a backbone link on application throghput", IEEE Network, 2002.
+
+
+2. Implementation of stateful protocols
+
+This reason is much less significant than the previous one.  We
+mention it here primarily so implementers of networking protocols are
+aware of the issue.  This issue is only relevant for a relatively
+small fraction of network protocols.
+
+Some networking protocols add sequence numbers to packets, and require
+either dropping packets that arrive out of sequence number order at a
+later network point (e.g. GRE with sequence numbers enabled), or with
+a looser check that allows some amount of network reordering to occur
+without dropping the packets (e.g. IPsec).  If a device implementing
```

Comment:
Nit: s/If/When/ 
and adjust the rest of the sentence?
"When a device implements these protocols and does the ...

Reply:
Done